porch/pkg/registry/porch/secret.go

```diff
@@ -44,7 +65,144 @@ func (r *secretResolver) ResolveCredential(ctx context.Context, namespace, name
+		}
+		if found {
+			return cred, nil
+		}
+	}
+	return repository.Credential{}, &NoMatchingResolverError{
+		Type: string(secret.Type),
+	}
+}
+
+type NoMatchingResolverError struct {
+	Type string
+}
+
+func (e *NoMatchingResolverError) Error() string {
+	return fmt.Sprintf("no resolver for secret with type %s", e.Type)
+}
+
+func (e *NoMatchingResolverError) Is(err error) bool {
+	nmre, ok := err.(*NoMatchingResolverError)
+	if !ok {
+		return false
+	}
+	return nmre.Type == e.Type
+}
+
+func NewBasicAuthResolver() Resolver {
+	return &BasicAuthResolver{}
+}
+
+var _ Resolver = &BasicAuthResolver{}
+
+type BasicAuthResolver struct{}
+
+func (b *BasicAuthResolver) Resolve(_ context.Context, secret core.Secret) (repository.Credential, bool, error) {
+	if secret.Type != BasicAuthType {
+		return repository.Credential{}, false, nil
+	}
+
 	return repository.Credential{
 		Data: secret.Data,
-	}, nil
+	}, true, nil
+}
+
+func NewGcloudWIResolver(corev1Client *corev1client.CoreV1Client, stsClient *stsv1.Service) Resolver {
+	return &GcloudWIResolver{
+		coreV1Client: corev1Client,
+		exchanger:    wi.NewWITokenExchanger(corev1Client, stsClient),
+		tokenCache:   make(map[tokenCacheKey]*oauth2.Token),
+	}
+}
+
+var _ Resolver = &GcloudWIResolver{}
+
+type GcloudWIResolver struct {
+	coreV1Client *corev1client.CoreV1Client
+	exchanger    *wi.WITokenExchanger
+
+	mutex      sync.Mutex
+	tokenCache map[tokenCacheKey]*oauth2.Token
+}
+
+type tokenCacheKey struct {
+	ksa types.NamespacedName
+	gsa string
+}
+
+var porchKSA = types.NamespacedName{
+	Name:      "porch-server",
+	Namespace: "porch-system",
+}
+
+func (g *GcloudWIResolver) Resolve(ctx context.Context, secret core.Secret) (repository.Credential, bool, error) {
+	if secret.Type != WorkloadIdentityAuthType {
+		return repository.Credential{}, false, nil
+	}
+
+	token, err := g.getToken(ctx)
+	if err != nil {
+		return repository.Credential{}, true, err
+	}
+	return repository.Credential{
+		Data: map[string][]byte{
+			"username": []byte("token"), // username doesn't matter here.
+			"password": []byte(token.AccessToken),
+		},
+	}, true, nil
+}
+
+func (g *GcloudWIResolver) getToken(ctx context.Context) (*oauth2.Token, error) {
+	gsa, err := g.lookupGSAEmail(ctx, porchKSA.Name, porchKSA.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenKey := tokenCacheKey{
+		ksa: porchKSA,
+		gsa: gsa,
+	}
+
+	g.mutex.Lock()
+	token, found := g.tokenCache[tokenKey]
+	g.mutex.Unlock()
+
+	if found {
+		timeLeft := time.Until(token.Expiry)
+		if timeLeft > 5*time.Minute {
+			return token, nil
+		}
+	}
+
+	token, err = g.exchanger.Exchange(ctx, porchKSA, gsa)
+	if err != nil {
+		return nil, err
+	}
+
+	g.mutex.Lock()
+	g.tokenCache[tokenKey] = token
+	g.mutex.Unlock()
+
+	return token, nil
+}
+
+func (g *GcloudWIResolver) lookupGSAEmail(ctx context.Context, name, namespace string) (string, error) {
+	sa, err := g.coreV1Client.ServiceAccounts(namespace).Get(ctx, name, v1.GetOptions{})
```

Comment:
We could just have the user specify the GCP ServiceAccount on the repository (instead of specifying a secret)